src/ZkFold/Base/Protocol/ARK/Plonk.hs

```diff
@@ -3,227 +3,157 @@
-        qo     = polyVecInLagrangeBasis @F @d @(PlonkMaxPolyDegree d) omega qO
-        qc     = polyVecInLagrangeBasis @F @d @(PlonkMaxPolyDegree d) omega qC
-        sigma1 = polyVecInLagrangeBasis @F @d @(PlonkMaxPolyDegree d) omega s1
-        sigma2 = polyVecInLagrangeBasis @F @d @(PlonkMaxPolyDegree d) omega s2
-        sigma3 = polyVecInLagrangeBasis @F @d @(PlonkMaxPolyDegree d) omega s3
-
-plonkVerifierInput :: Vector n F -> PlonkInput
+        qm     = polyVecInLagrangeBasis @(ScalarField c1) @n @(PlonkPolyExtendedLength n) omega qM
+        ql     = polyVecInLagrangeBasis @(ScalarField c1) @n @(PlonkPolyExtendedLength n) omega qL
+        qr     = polyVecInLagrangeBasis @(ScalarField c1) @n @(PlonkPolyExtendedLength n) omega qR
+        qo     = polyVecInLagrangeBasis @(ScalarField c1) @n @(PlonkPolyExtendedLength n) omega qO
+        qc     = polyVecInLagrangeBasis @(ScalarField c1) @n @(PlonkPolyExtendedLength n) omega qC
+        sigma1 = polyVecInLagrangeBasis @(ScalarField c1) @n @(PlonkPolyExtendedLength n) omega s1
+        sigma2 = polyVecInLagrangeBasis @(ScalarField c1) @n @(PlonkPolyExtendedLength n) omega s2
+        sigma3 = polyVecInLagrangeBasis @(ScalarField c1) @n @(PlonkPolyExtendedLength n) omega s3
+
+plonkVerifierInput :: Field (ScalarField c) => Vector n (ScalarField c) -> PlonkInput c
 plonkVerifierInput input = PlonkInput $ fromList $ map negate $ fromVector input
 
-instance forall d n t .
-        (KnownNat d,
-         KnownNat n,
-         KnownNat (PlonkPermutationSize d),
-         KnownNat (PlonkMaxPolyDegree d),
-         ToTranscript t F,
-         ToTranscript t G1,
-         FromTranscript t F) => NonInteractiveProof (Plonk d n t) where
-    type Transcript (Plonk d n t)  = t
-    type SetupProve (Plonk d n t)  = (PlonkSetupParamsProve, PlonkPermutation d, PlonkCircuitPolynomials d, PlonkWitnessMap d)
-    type SetupVerify (Plonk d n t) = (PlonkSetupParamsVerify, PlonkCircuitCommitments)
-    type Witness (Plonk d n t)     = (PlonkWitnessInput, PlonkProverSecret)
-    type Input (Plonk d n t)       = PlonkInput
-    type Proof (Plonk d n t)       = PlonkProof
-
-    setupProve :: Plonk d n t -> SetupProve (Plonk d n t)
+instance forall n l c1 c2 t plonk f g1.
+        ( Plonk n l c1 c2 t ~ plonk
+        , ScalarField c1 ~ f
+        , Point c1 ~ g1
+        , KnownNat n
+        , KnownNat l
+        , KnownNat (PlonkPermutationSize n)
+        , KnownNat (PlonkPolyExtendedLength n)
+        , Eq (ScalarField c1)
+        , Scale (ScalarField c1) (ScalarField c1)
+        , BinaryExpansion (ScalarField c1)
+        , Bits (ScalarField c1) ~ [ScalarField c1]
+        , FiniteField (ScalarField c1)
+        , AdditiveGroup (BaseField c1)
+        , Pairing c1 c2
+        , ToTranscript t (ScalarField c1)
+        , ToTranscript t (Point c1)
+        , FromTranscript t (ScalarField c1)
+        ) => NonInteractiveProof (Plonk n l c1 c2 t) where
+    type Transcript (Plonk n l c1 c2 t)  = t
```

Comment:
replace `Plonk n l c1 c2 t` to `plonk` ?

Reply:
Unfortunately, you cannot replace it here. Only in class methods.